internal/tool/tool.go

```diff
@@ -30,27 +35,195 @@ var _ codacy.Tool = (*codacySemgrep)(nil)
+	configFile, err = createConfigFile(toolExecution.SourceDir, toolExecution.Patterns)
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll("tmp/")
+	defer configFile.Close()
 
-	semgrepCmd := semgrepCommand(configFile, *toolExecution.Files, toolExecution.SourceDir)
+	err = populateFilesByLanguage(toolExecution.Files, toolExecution.SourceDir)
+	if err != nil {
+		return nil, errors.New("Error getting files to analyse: " + err.Error())
+	}
 
-	semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+	result, err := run(configFile, toolExecution)
 	if err != nil {
-		return nil, errors.New("Error running semgrep: " + semgrepError)
+		return nil, err
 	}
 
-	result := parseOutput(toolExecution.ToolDefinition, semgrepOutput)
 	return result, nil
 }
 
+func run(configFile *os.File, toolExecution codacy.ToolExecution) ([]codacy.Result, error) {
+	var result []codacy.Result
+	for language, files := range filesByLanguage {
+		semgrepCmd := semgrepCommand(configFile, toolExecution.SourceDir, language, files)
+
+		semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+		if err != nil {
+			return nil, errors.New("Error running semgrep: " + semgrepError)
+		}
+
+		output, err := parseOutput(toolExecution.ToolDefinition, semgrepOutput)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, output...)
+	}
+
+	return result, nil
+}
+
+func populateFilesByLanguage(toolExecutionFiles *[]string, toolExecutionSourceDir string) error {
+	// If there are files to analyse, analyse only those files
+	if toolExecutionFiles != nil && len(*toolExecutionFiles) > 0 {
+		return populateFilesByLanguageFromFiles(*toolExecutionFiles)
+	}
+	// If there are no files to analyse, analyse all files from source dir
+	return populateFilesByLanguageFromSourceDir(toolExecutionSourceDir)
+}
+
+func populateFilesByLanguageFromFiles(toolExecutionFiles []string) error {
+	for _, file := range toolExecutionFiles {
+		addFileToFilesByLanguage(file)
+	}
+
+	return nil
+}
+
+func populateFilesByLanguageFromSourceDir(toolExecutionSourceDir string) error {
+	// Semgrep can analyse full directories and its subdirectories
+	// but we will have to analyse every extension from every file
+	// so we will have to do this walk somewhere else if we dont do it here
+	err := filepath.WalkDir(toolExecutionSourceDir, func(path string, info fs.DirEntry, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+		pathInfo, pathErr := info.Info()
+		if pathErr != nil {
+			return pathErr
+		}
+		// if it is a file and it is not a hidden file
+		if !pathInfo.IsDir() && !strings.HasPrefix(pathInfo.Name(), ".") {
+			addFileToFilesByLanguage(path)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addFileToFilesByLanguage(fileName string) {
+	language := detectLanguage(fileName)
+	filesByLanguage[language] = append(filesByLanguage[language], fileName)
+}
+
+// This feels illegal
+func detectLanguage(fileName string) string {
```

Comment:
> we should aim to use the same extensions as defined here: https://github.com/codacy/codacy-plugins-api/blob/e94cfa10a5f2eafdeeeb91e30a39e2032e1e4cc7/codacy-plugins-api/src/main/scala/com/codacy/plugins/api/languages/Language.scala#L100

this list contains all languages which semgrep can detect. the idea here would be to change (or add) any missing extension based on the list you provided correct?

Reply:
yes 👍 
we should try to guarantee that at least the file extensions from the `Supported` languages are covered in this function